code/modules/power/supermatter/supermatter.dm

```diff
@@ -397,26 +485,251 @@ GLOBAL_DATUM(main_supermatter_engine, /obj/machinery/power/supermatter_crystal)
+ * [/obj/machinery/power/supermatter_crystal/var/gas_heat_power_generation]
+ * [/obj/machinery/power/supermatter_crystal/var/gas_powerloss_inhibition]
+ *
+ * Returns: null
+ */
+/obj/machinery/power/supermatter_crystal/proc/calculate_gases()
+	if(disable_gas)
+		return
+	gas_percentage = list()
+	gas_power_transmission = 0
+	gas_heat_modifier = 0
+	gas_heat_resistance = 0
+	gas_heat_power_generation = 0
+	gas_powerloss_inhibition = 0
+
+	var/total_moles = absorbed_gasmix.total_moles()
+
+	for (var/gas_path in absorbed_gasmix.gases)
+		gas_percentage[gas_path] = absorbed_gasmix.gases[gas_path][MOLES] / total_moles
+		var/datum/sm_gas/sm_gas = GLOB.sm_gas_behavior[gas_path]
+		if(!sm_gas)
+			continue
+		gas_power_transmission += sm_gas.power_transmission * gas_percentage[gas_path]
+		gas_heat_modifier += sm_gas.heat_modifier * gas_percentage[gas_path]
+		gas_heat_resistance += sm_gas.heat_resistance * gas_percentage[gas_path]
+		gas_heat_power_generation += sm_gas.heat_power_generation * gas_percentage[gas_path]
+		gas_powerloss_inhibition += sm_gas.powerloss_inhibition * gas_percentage[gas_path]
+
+	gas_heat_power_generation = clamp(gas_heat_power_generation, 0, 1)
+	gas_powerloss_inhibition = clamp(gas_powerloss_inhibition, 0, 1)
+
+/**
+ * Perform calculation for power lost and gained this tick.
+ * Description of each factors can be found in the defines.
+ *
+ * Updates:
+ * [/obj/machinery/power/supermatter_crystal/var/power]
+ * [/obj/machinery/power/supermatter_crystal/var/external_power_trickle]
+ *
+ * Returns: The factors that have influenced the calculation. list[FACTOR_DEFINE] = number
+ */
+/obj/machinery/power/supermatter_crystal/proc/calculate_internal_power()
+	if(disable_power_change)
+		return
+	var/list/additive_power = list()
+
+	/// If we have a small amount of external_power_trickle we just round it up to 40.
+	additive_power[SM_POWER_EXTERNAL_TRICKLE] = external_power_trickle ? max(external_power_trickle/MATTER_POWER_CONVERSION, 40) : 0
+	external_power_trickle -= min(additive_power[SM_POWER_EXTERNAL_TRICKLE], external_power_trickle)
+	additive_power[SM_POWER_EXTERNAL_IMMEDIATE] = external_power_immediate
+	external_power_immediate = 0
+	additive_power[SM_POWER_HEAT] = gas_heat_power_generation * absorbed_gasmix.temperature / 6
+
+	// I'm sorry for this, but we need to calculate power lost immediately after power gain.
+	// Helps us prevent cases when someone dumps superhothotgas into the SM and shoots the power to the moon for one tick.
+	/// Power if we dont have decay. Used for powerloss calc.
+	var/momentary_power = power
+	for(var/powergain_type in additive_power)
+		momentary_power += additive_power[powergain_type]
+	if(power < powerloss_linear_threshold) // Negative numbers
+		additive_power[SM_POWER_POWERLOSS] = -1 * (momentary_power / POWERLOSS_CUBIC_DIVISOR) ** 3
+	else
+		additive_power[SM_POWER_POWERLOSS] = -1 * momentary_power * POWERLOSS_LINEAR_RATE + powerloss_linear_offset
```

Comment:
```suggestion
		additive_power[SM_POWER_POWERLOSS] = -1 * (momentary_power * POWERLOSS_LINEAR_RATE + powerloss_linear_offset)
```

The offset should get added to the product of powerloss and the poqerloss linear rate.